feat: SHA-256 checksum

dkorbelainen opened the pull request:
- Implement public domain SHA-256 algorithm
- Add checksum calculation in allocator for block verification
---
- Automatically calculate SHA-256 on block write
- Automatically verify SHA-256 on block read
---
- Implement calculate_checksum() and verify_checksum() methods
- Log warning when checksum verification fails
---
- Checksum tests

dkorbelainen committed:
feat: add SHA-256 checksum for data integrity verification

- Implement public domain SHA-256 algorithm
- Add checksum calculation in allocator for block verification

dkorbelainen committed:
feat: integrate SHA-256 checksum verification into storage blocks
- Add checksum field to storage_block structure
- Automatically calculate SHA-256 on block write
- Automatically verify SHA-256 on block read
- Update STORAGE_BLOCK_METASIZE (+64 bytes for checksum)
- Implement calculate_checksum() and verify_checksum() methods
- Log warning when checksum verification fails
Overhead impact: +64 bytes per block (~1.5-2% for typical 4KB blocks)

dkorbelainen committed:
test: add comprehensive data integrity tests with SHA-256 verification

- Test basic checksum verification on write/read
- Test detection of data corruption
- Test integrity with multiple files
- Test integrity with large files (100KB+)
- Test edge cases: empty files and overwrites

mozhaa commented on include/compio/file.hpp:
```diff
@@ -130,6 +130,7 @@ struct storage_block : public infile_object {
     uint64_t size;                   /**< Size of data array */
     uint64_t original_size;          /**< Original size (size of uncompressed data) */
     std::unique_ptr<uint8_t[]> data; /**< Data block */
+    char checksum[65];               /**< SHA-256 checksum (64 hex chars + null terminator) */
```
sha-256 вроде как 32 байта должен занимать
ну и нуль терминатор не нужен, раз уж тут константный размер (ты вроде его и не читаешь из файла даже)

dkorbelainen replied:
В шестнадцатиричном формате байт это 2 символа, поэтому 64. 
Насчет нуль-терминатора согласен

mozhaa commented on src/file.cpp:
```diff
@@ -256,4 +275,25 @@ int files_table::remove(const char *name) {
+        return;
+    }
+
+    std::string hash = SHA256::compute(data.get(), size);
```
может использовать fnv1a хэш, который уже реализован в utils.cpp? как будто раз у нас блоки небольшие, то 32 байта на каждый блок тратить для хэша это перебор. fnv1a - 4 байта если что
так-то самая идея чексуммы хорошая

mozhaa replied:
хотя не знаю, может и нормально 32 байта... можно с Никитой Юрьевичем обсудить

dkorbelainen replied:
Я вообще при анализе тоже скептически смотрел на такую серьезную добавку к блоку, но там дело в том что оверхед самый серьезный для самых маленьких блоков (1Кб т д), чем больше тем незаметнее становится. На стандартных блоках 4Кб оверхед 0.5%->2% (если 16Кб, 64Кб то вообще 0.4%, 0.1% соответственно), что по идее должно быть приемлемо для защиты от повреждений, с одной стороны.

С другой стороны, если менять на FNV-1 то экономия 52 байт на блок.
Типа SHA-256 overhead 64/4096 = 1.56%
FNV-1a overhead 8/4096 = 0.2% и 4/4096 = 0.1%
Но в целом, контрольные суммы так-то нужны для детекции случайных повреждений, а не чего-то более серьезного вроде злонамеренных атак (ну как минимум сейчас?), поэтому с этой стороны FNV-1a достаточно. И видимо более разумно, так как он в 8 раз меньше по оверхеду и быстрее

В общем пока что вынес варианты в отдельный файл  (`/src/checksum_alternatives.cpp`), пусть будет. SHA256 криптографически стойкий, но медленный, возможно пригодится, ну или хотя бы для анализа и сравнения, чтоб было что сказать на защите по этому поводу, ну и 2 версии FNV-1a 32/64-bit. 
Поменял на FNV-1a 32bit.

mozhaa reviewed (changes requested):
вроде все норм, кроме того, что написал

dkorbelainen committed:
refactor: replace SHA-256 with FNV-1a 32-bit for checksum

- Replace 64-byte SHA-256 checksum with 4-byte FNV-1a 32-bit hash
- Reduce overhead from 1.56% to 0.10% for 4KB blocks
- Add fnv1a_32() implementation for binary data in utils.cpp
- Update STORAGE_BLOCK_METASIZE to reflect new checksum size
- Preserve SHA-256 implementation in checksum_alternatives.cpp for future comparison

FNV-1a 32-bit provides sufficient protection against accidental data
corruption while being significantly faster and more space-efficient
than cryptographic hashes.
